File: app/src/main/java/de/robinrehbein/punkt/game/MedalTier.kt
package de.robinrehbein.punkt.game

import androidx.annotation.StringRes
import de.robinrehbein.punkt.R

/**
 * Medaillen-Stufen ab 10/20/30/40 Punkten. Die Reihenfolge der Einträge
 * ist gleichzeitig die Rangfolge (ordinal) — daran hängt die
 * "NEUE MEDAILLE!"-Feier im Game-Over.
 *
 * Schwellen und Farben stehen nicht mehr hier, sondern in [MedalPaint]
 * im :core-Modul — dieselbe Quelle nutzt auch die Uhr. Diese Aufzählung
 * verbindet sie nur noch mit den String-Ressourcen der App.
 */
enum class MedalTier(val id: MedalId, @StringRes val nameRes: Int) {
    BRONZE(MedalId.BRONZE, R.string.medal_bronze),
    SILVER(MedalId.SILVER, R.string.medal_silver),
    GOLD(MedalId.GOLD, R.string.medal_gold),
    PLATINUM(MedalId.PLATINUM, R.string.medal_platinum);

    val threshold: Int get() = MedalPaint.threshold(id)

    /** Körperfarbe der Münze als ARGB-Long (siehe medalColors). */
    val body: Long get() = MedalPaint.body(id)

    /** Schatten- und Prägefarbe der Münze als ARGB-Long. */
    val shade: Long get() = MedalPaint.shade(id)

    companion object {
        private fun of(id: MedalId?): MedalTier? =
            id?.let { wanted -> entries.first { it.id == wanted } }

        /** Höchste erreichte Stufe, null unterhalb von Bronze. */
        fun forScore(score: Int): MedalTier? = of(MedalPaint.forScore(score))

        /** Nächste noch nicht erreichte Stufe, null ab Platin. */
        fun next(score: Int): MedalTier? = of(MedalPaint.next(score))

        /** Bringt dieser Score eine höhere Stufe als der bisherige Bestwert? */
        fun isUpgrade(score: Int, previousBest: Int): Boolean =
            MedalPaint.isUpgrade(score, previousBest)
    }
}

File: ios/Dottie/Sources/Engine/MedalTier.swift
import Foundation

/// Port von app/.../MedalTier.kt: Medaillen-Stufen ab 10/20/30/40 Punkten.
/// Die Reihenfolge der Fälle ist gleichzeitig die Rangfolge (rawValue) —
/// daran hängt die "NEUE MEDAILLE!"-Feier im Game-Over.
enum MedalTier: Int, CaseIterable {
    case bronze
    case silver
    case gold
    case platinum

    var threshold: Int {
        switch self {
        case .bronze: return 10
        case .silver: return 20
        case .gold: return 30
        case .platinum: return 40
        }
    }

    /// Localizable.strings-Key des Namens.
    var nameKey: String {
        switch self {
        case .bronze: return "medal_bronze"
        case .silver: return "medal_silver"
        case .gold: return "medal_gold"
        case .platinum: return "medal_platinum"
        }
    }

    /// Höchste erreichte Stufe, nil unterhalb von Bronze.
    static func forScore(_ score: Int) -> MedalTier? {
        return allCases.last(where: { score >= $0.threshold })
    }

    /// Nächste noch nicht erreichte Stufe, nil ab Platin.
    static func next(_ score: Int) -> MedalTier? {
        return allCases.first(where: { score < $0.threshold })
    }

    /// Bringt dieser Score eine höhere Stufe als der bisherige Bestwert?
    static func isUpgrade(score: Int, previousBest: Int) -> Bool {
        let newRank = forScore(score)?.rawValue ?? -1
        let oldRank = forScore(previousBest)?.rawValue ?? -1
        return newRank > oldRank
    }
}
